woocommerce/woocommerce-ios
# [Mobile Payments] Add model objects with metadata necessary to print receipts

Closes #3936 

⚠️ This PR is against the feature branch implementing support for Card Present Payments, not against develop ⚠️
⚠️ Testing this PR requires a hardware reader ⚠️

We continue building our prototype (for reference, p91TBi-4q4-p2), this time, by adding the fields needed to print a receipt.

Receipts can be customized, as long as we list some required information. When accepting in-person payments with EMV chip cards, card networks require us to include several fields on the receipts we provide to customers. Those fields are inluded as part of the PaymentIntent object (to be more specific, the CardReaderDetails object), once the payment is processed. For more information, see Stripe's documentation: [Custom receipts](https://stripe.com/docs/terminal/checkout/receipts#custom)

So this PR added the necessary fields to `Charge`, so that [Charges](https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPCharge.html) associated to a [Payment Intent](https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPPaymentIntent.html) contain the information needed to print a receipt. Sorry it went over 500 lines, but in between docs, changes to the project files, headerdocs and tests, it has ballooned a bit, and it was not pragmatic to break it down further.

## Changes
The PR adds the following model objects:
* PaymentMethod. An enumeration that models the payment method associated with the charge. This enumeration can have an associated value for one of its cases:
* CardPresentDetails. The value associated to the case .cardPresent of PaymentMethod. This objects represents the details from a transaction using a card_present payment method. Among other properties, it contains:
* ReceiptDetails. Receipt details associated with a card present transaction. This is the data that we actually need to render into the receipt. 
* CardBrand. An enumeration to model the type of card. 
* Extension to all the mentioned model objects, to provide initializers that map between the model objects provided by Stripe (which we do not want to expose to the rest of the codebase) an our domain objects.
* Some unit tests to cover those mappings, when possible.

Note how there is not a 1:1 correspondence between our model objects and Stripe's. Some objects the Stripe SDK make sense in a context where said objects need to be consumed from Obj-C code (e.g. SCPaymentMethod and SCPaymentMethodType), while for us, being all Swift, there are better ways to model that data (e.g. hardware.PaymentMethod)

## How to test
Testing is a bit complicated because:
1. Requires setting up a store for [Card Present Payment Testing](https://github.com/woocommerce/woocommerce-ios/blob/issue/3748-intent-service/docs/stripe-tests.md#integration-tests) (also P91TBi-4BH-p2 for more specific instructions)  
2. Requires an external hardware reader.

* Checkout the branch, run `bundle exec pod install`, just for kicks and giggles.
* **Add break point to line 113 of CardPresentPaymentStore** :

<img width="350" alt="Screen Shot 2021-04-19 at 13 38 24" src="https://user-images.githubusercontent.com/2722505/115279384-ad122f00-a114-11eb-8dd1-6f5926c2681b.png">


* Build and run on a device. 
* Switch an external card reader on. The only card reader supported by now is the BBPOS Chipper 2X BT.
* Log in to an account that matches a store that has been configured for testing.
* Navigate to Settings in the app. (Tap the gear button in the top right)
* In the "Store Settings" section, tap "Manage Card Reader".
* Make sure the card reader is on and the blue light is blinking (we don't do any error handling yet, so it's important that things are setup for the happy path)
* Tap "Connect card reader"
* Wait until the app suggests one reader to connect to.
* Tap "Connect to reader"
* Make sure the reader light turns solid blue.
* Navigate to an order eligible for Card Present Payments (it has to be a store enrolled to WCPay, and the order needs to have as payment method associated  Cash on Delivery or none).
* Tap the "Collect Payment" button.
* Tap the card to the reader when requested.
* If everything goes well, the payment should be processed and the breakpoint added to CardPresentPaymentStore should be triggered.
* `po intent`

The resulting instance of PaymentIntent returned by the card reader service after the payment has been processed should look like this:

```
▿ PaymentIntent
  - id : "pi_REDACTED"
  - status : Hardware.PaymentIntentStatus.requiresCapture
  ▿ created : 2021-04-19 17:38:05 +0000
    - timeIntervalSinceReferenceDate : 640546685.0
  - amount : 21000
  - currency : "usd"
  ▿ metadata : Optional<Dictionary<AnyHashable, Any>>
    - some : 0 elements
  ▿ charges : 1 element
    ▿ 0 : Charge
      - id : "ch_REDACTED"
      - amount : 21000
      - currency : "usd"
      - status : Hardware.ChargeStatus.succeeded
      ▿ description : Optional<String>
        - some : "Receipt description."
      ▿ metadata : Optional<Dictionary<AnyHashable, Any>>
        - some : 0 elements
      ▿ paymentMethod : Optional<PaymentMethod>
        ▿ some : PaymentMethod
          ▿ presentCard : 1 element
            ▿ details : CardPresentDetails
              - last4 : "9999"
              - expMonth : 12
              - expYear : 2021
              ▿ cardholderName : Optional<String>
                - some : "TEST/STRIPE"
              - brand : Hardware.CardBrand.visa
              - fingerprint : "REDACTED"
              ▿ generatedCard : Optional<String>
                - some : "pm_REDACTED"
              ▿ receipt : Optional<ReceiptDetails>
                ▿ some : ReceiptDetails
                  - applicationPreferredName : "Stripe Credit"
                  - dedicatedFileName : "A000000003101001"
                  - authorizationResponseCode : "3030"
                  - applicationCryptogram : "36E4873E616952E7"
                  - terminalVerificationResults : "0000000000"
                  - transactionStatusInformation : "0000"
                  ▿ accountType : Optional<String>
                    - some : "credit"
              ▿ emvAuthData : Optional<String>
                - some : "8A023030"
```

Notice the object tree structure: PaymentIntent > Charge > PaymentMethod > CardPresentDetails > ReceiptDetails

* Also, CI should be ✅ 

Update release notes:
